Migrate to aiohttp and TCP/IP from urllib and unix sockets

aleksul opened the pull request:
Resolves #17.

Changes:
- plugin.py
  - binary download fixture is now using aiohttp.
  - Unix socket fixtures were replaced by new_port fixture, along with mongod binding to localhost instead of socket.
  - Additionally, I disabled mongod --fork option, because we already use asyncio.create_subprocess_exec which takes care of the process. See [mongod docs](https://docs.mongodb.com/manual/reference/program/mongod/#std-option-mongod.--fork). It might be the case of issue #23, need to check twice.
- Add aiohttp in requirements.txt.
- I didn't write new tests, because yours seem to already cover these changes.

aleksul committed:
Migrate from urllib to aiohttp
Migrate from UNIX Socket to TCP/IP

aleksul committed:
Merge branch 'AustinScola:master' into master

AustinScola commented on pytest_motor/plugin.py:
```diff
@@ -3,10 +3,11 @@
-import urllib.request
 from pathlib import Path
 from typing import AsyncIterator, Iterator, List
+import socket
```
You can use `scripts/format.sh` to format code which includes sorting imports. See [`CONTRIBUTING.md#scripts`](https://github.com/AustinScola/pytest-motor/blob/master/CONTRIBUTING.md#scripts) which I just added recently.

aleksul replied:
Sorry, didn't see that. Will fix in a minute

AustinScola commented on pytest_motor/plugin.py:
```diff
@@ -51,45 +52,36 @@ async def root_directory(pytestconfig: PytestConfig) -> Path:
-    mongod_binary_filename = destination.joinpath(
-        'mongodb-linux-x86_64-ubuntu1804-4.4.6/bin/mongod')
+    mongod_binary_relative_path = 'mongodb-linux-x86_64-ubuntu1804-4.4.6/bin/mongod'
+    mongod_binary_filename = destination.joinpath(mongod_binary_relative_path)
```
Nit: might be worth renaming `mongod_binary_filename` to `mongod_binary_path` now too?

AustinScola commented on pytest_motor/plugin.py:
```diff
@@ -117,19 +109,23 @@ async def database_path(databases_directory: Path) -> AsyncIterator[Path]:
-        str(unix_socket), '--storageEngine', 'ephemeralForTest', '--fork', '--logpath', '/dev/null',
-        '--dbpath',
-        str(database_path)
+        str(mongod_binary),
```
This looks much prettier with `yapf` disabled, thanks!

AustinScola commented on pytest_motor/plugin.py:
```diff
@@ -51,45 +52,36 @@ async def root_directory(pytestconfig: PytestConfig) -> Path:
+        async with aiohttp.ClientSession() as session:
+            async with session.get(download_url) as resp:
+                with open('binaries.tar', 'wb') as binary_file:
+                    # Read by chunks to avoid big RAM consumption
```
This is cool, I wouldn't have thought of doing that.

aleksul replied:
Mainly, this is aiohttp snippet. Another approach is to use aiofiles library, but this will add even more unnecessary dependencies. I'm not 100% sure, but maybe the best way is to use tempfile lib.

AustinScola commented on pytest_motor/plugin.py:
```diff
@@ -51,45 +52,36 @@ async def root_directory(pytestconfig: PytestConfig) -> Path:
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as opened_socket:
+        opened_socket.bind(("127.0.0.1", 0))  # system will automaticly assign port
+        port = opened_socket.getsockname()[1]
+    return port
```
I think we probably want to yield the port inside of the context manager. According to the [socket object docs](https://docs.python.org/3/library/socket.html#socket-objects), "Exiting the context manager is equivalent to calling `close()`".

aleksul replied:
I don't think that's a good idea. If I understand correctly, this means that python will reserve this socket, and mongod won't bind it, because it's already allocated.

AustinScola replied:
Hmm okay. So we probably don't want to do that then, but but should we still perform some cleanup afterwards?

aleksul replied:
What do you mean by cleanup? Mongod will free the port as it closes, won't it?

AustinScola replied:
I wasn't sure if it does or not?

aleksul replied:
I think it does. Will write a test for it.

aleksul replied:
Added this test and it passes! 🎉 

```python
@pytest.mark.asyncio
async def test_port_not_allocated(mongod_binary: Path, new_port: int) -> None:
    """Tests that mongod leaves port free"""
    mongod = await asyncio.create_subprocess_exec(*[str(mongod_binary), '--port', str(new_port)])
    mongod.terminate()
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as opened_socket:
        assert opened_socket.connect_ex(('127.0.0.1', new_port)) != 0
```

aleksul commented:
Oops... looks like there are more requirements files to be updated

AustinScola commented on requirements/requirements.txt:
```diff
@@ -1,3 +1,4 @@
 motor==2.4.0
 pytest==6.2.4
 pytest-asyncio==0.15.1 # Needed for async pytest fixtures
+aiohttp==3.7.4.post0
```
The way that I currently have `pytest-motor` set up is that the scripts use frozen requirements files in `requirements/frozen`. To update the frozen requirements files based on changes to the requirements files, you can run `scripts/freeze.sh`. Also, aiohttp should be added to `install_requires` in `setup.py`.

aleksul committed:
Update frozen requirements

AustinScola reviewed (commented):
@aleksul thanks for addressing this so quickly. I just have some minor suggestions for the PR. Long term, I think I would like to maintain support for unix sockets in addition to ports (I think unix sockets are faster than ports?), but that can come later. I'm fine with ports being the default connection mechanism for now.

> I disabled mongod --fork option, because we already use asyncio.create_subprocess_exec which takes care of the process

Good catch!

> I didn't write new tests, because yours seem to already cover these changes.

The testing situation is a bit misleading right now. Full coverage is being reported but that is because `test_motor_client` test is more like an integration test. I would prefer to have individual unit tests for everything eventually but I'll leave it up to you whether you want to implement them here or not

AustinScola commented:
> Oops... looks like there are more requirements files to be updated

Yes, see my comment on the requirements file :)

aleksul committed:
Format imports

aleksul committed:
Update requirements one more time :)

aleksul committed:
Change mongod binary path naming

aleksul commented:
> The testing situation is a bit misleading right now. Full coverage is being reported but that is because `test_motor_client` test is more like an integration test. I would prefer to have individual unit tests for everything eventually but I'll leave it up to you whether you want to implement them here or not

This doesn't seem too hard for me. Will do it within next few hours.

aleksul committed:
Saving .tar as a tempfile

aleksul committed:
Add tests for port and binary_downloader
Improved test_plugin a little bit
Renamed multiple_tests to independence_tests
Separated independece_tests from motor_client_test

aleksul commented:
Test coverage script reports 97% on my machine... This might be okay, for now? 👉👈

AustinScola commented on test_data/files/server_info_test.py:
```diff
@@ -6,4 +6,4 @@
 async def test_using_motor_client_server_info(motor_client: AsyncIOMotorClient) -> None:
     """Test retrieving the Motor client server info."""
-    await motor_client.server_info()
+    assert await motor_client.server_info() is not None
```
Can `server_info` return `None`?

aleksul replied:
I suppose it's better to assert if it equals to some string. Any suggestions?

AustinScola replied:
Probably? The test was really just meant to be the simplest "does it work" (and by that I mean connect) test I could write. I would think the `server_info` would be dynamic.

aleksul replied:
I found out that server_info will throw an exception if it can't connect via motor_client. So I rewrote the test to assert False + message if connection timeout exception was thrown. I think it adds a little more clarity to this test.

AustinScola commented:
> Test coverage script reports 97% on my machine... This might be okay, for now?

Want to just mark it with `# pragma: no cover`? I think it is a pretty safe part to ignore.

AustinScola commented:
After that, are there any other changes you want to make? If not we can merge :)

aleksul committed:
Update server info test

aleksul commented:
> 
> 
> > Test coverage script reports 97% on my machine... This might be okay, for now?
> 
> Want to just mark it with `# pragma: no cover`? I think it is a pretty safe part to ignore.

I want to, the only problem I can't really understand what are these uncovered 3%. Is there any command to see it?

aleksul committed:
Add #no-cover for mongod.terminate except block

aleksul commented:
> 
> 
> > > Test coverage script reports 97% on my machine... This might be okay, for now?
> > 
> > 
> > Want to just mark it with `# pragma: no cover`? I think it is a pretty safe part to ignore.
> 
> I want to, the only problem I can't really understand what are these uncovered 3%. Is there any command to see it?

Answering myself - `pytest --cov-report term-missing --cov=pytest_motor tests/` does the trick.

aleksul commented:
> 
> 
> After that, are there any other changes you want to make? If not we can merge :)

I think that's it for now 🎉

AustinScola commented:
> Answering myself - `pytest --cov-report term-missing --cov=pytest_motor tests/` does the trick.

Just saw this, the script `scripts/coverage.sh` can be used too.

AustinScola commented:
`pymongo` isn't typed. We can probably just add the following to the `mypy.ini` file for now:
```ini
[mypy-pymongo.*]
ignore_missing_imports = True
```

To check the type checking locally you can use `scripts/type_check.sh`

aleksul committed:
Add pymongo import to mypy ignores

aleksul commented:
> 
> 
> `pymongo` isn't typed. We can probably just add the following to the `mypy.ini` file for now:
> 
> ```ini
> [mypy-pymongo.*]
> ignore_missing_imports = True
> ```
> 
> To check the type checking locally you can use `scripts/type_check.sh`

Done!

aleksul commented:
This isn't dedicated to this PR, but I would suggest a small change in mypy script
from
`python3 -m mypy .`
to
`python3 -m mypy ./pytest_motor ./test_data ./tests`
because otherwise mypy scans venvs directory.

AustinScola commented:
> This isn't dedicated to this PR, but I would suggest a small change in mypy script
> from
> `python3 -m mypy .`
> to
> `python3 -m mypy ./pytest_motor ./test_data ./tests`
> because otherwise mypy scans venvs directory.

The `mypy.ini` has an `exclude` line to prevent that:
```ini
exclude=/venvs/
```

And generally, I prefer excluding files to check rather than including because it is very easy to add new files and not realize that `mypy` isn't checking them if you list includes.
